### PR TITLE
Support vision models and function calling

### PR DESCRIPTION
**Pull request by DePasqualeOrg**
I've added functionality from the [huggingface.js](https://github.com/huggingface/huggingface.js/tree/main/packages/jinja) implementation. This is a work in progress.

### CODE REVIEW

**Comment by johnmai-dev** [comment]
Thank you for your PR, it's great! Could you please provide some test cases?

I may not have time until next month. I've been a bit busy lately. 

Hi @pcuenca ! Do you have time to help review this PR?

**Comment by DePasqualeOrg** [comment]
I'll add some tests for images and function calling and try to polish this up a bit.

**Comment by pcuenca** [comment]
Hi @DePasqualeOrg, thanks a lot for the effort! It's a long diff, I can try to take a look in a couple of days. Do we need everything at once, including namespaces, built-in functions and tool calling, or could this potentially be approached in a few phases?

**Comment by DePasqualeOrg** [comment]
Tool calling could definitely be postponed for later. That seems to be more complicated to test for anyway. Since we now have support for vision language models in mlx-libraries, maybe that should be the focus here.

If it's too hard to separate out the changes, maybe we can focus on test coverage for image models in this PR, and add tests for function calling in a later PR?

**Comment by DePasqualeOrg** [comment]
It looks like Qwen2-VL is the only other major vision language model that includes image handling in the chat template, so now we have tests for image handling in Qwen2-VL and Llama 3.2, and I think this is ready for review.

Sorry in advance for the somewhat messy approach here. I did actually spend the better part of a day putting everything together. I'm sure it's not perfect, but I wanted to get the ball rolling on this.

One thought I had was that we can benefit from the work that has already been done on the TypeScript implementation by keeping the Swift implementation as close to it as possible. As LLMs get even better, they'll be able to port things more easily, and keeping the libraries roughly isomorphic would help with that.

**Comment by DePasqualeOrg** [comment]
I've rebased this after formatting the repo with swift-format, to make it easier for @pcuenca to review it. The tests for images and videos are passing. For now I'm not testing the tool-calling functionality. My goal is to get the chat templates for vision models working and to focus on tool calling later.

**Comment by DePasqualeOrg** [comment]
The existing tests as well as some additional ones I added pass. I added some tests for tool use based on the ones in the TypeScript implementation, but I can't get them to pass after many hours of trying. Someone smarter than me will have to take it from here. This should already work for vision models, though.

**Comment by DePasqualeOrg** [comment]
Four of the six tool use tests are now passing. It's quite difficult to get these tests to pass, because dictionaries in Swift don't preserve the order of the keys. I've used OrderedDictionary in the tests, which presents its own challenges, since sometimes these values need to be converted to JSON. I think this should work in practice for many models that use function calling. Perhaps we should make tests for the more recent ones, such as Llama 3.2.

**Comment by DePasqualeOrg** [comment]
This adds quite a lot of functionality from Jinja2 in Python. I've added tests from the TypeScript Jinja library and from Jinja2. Some of the new tests I've added from Jinja2 aren't passing and are disabled, but they mostly seem to cover edge cases. I also created tests for vision language models and tool use.

I could keep going with this, but I think this is a good place to stop for now, since it seems to work with chat templates for vision language models and tool use. Maybe I'll polish it up a bit later on, but I've already spent several whole days on this and need to move on to other things.

@pcuenca and @johnmai-dev, I don't expect you to review every line of this, but maybe you can scan through it and let me know if anything catches your eye.

**Comment by DePasqualeOrg** [comment]
@pcuenca and @johnmai-dev, if you don't have any feedback, I think we should merge this so that I can move forward with https://github.com/huggingface/swift-transformers/pull/151 and https://github.com/ml-explore/mlx-swift-examples/pull/173.

**Comment by johnmai-dev** [comment]
> @pcuenca and @johnmai-dev, if you don't have any feedback, I think we should merge this so that I can move forward with [huggingface/swift-transformers#151](https://github.com/huggingface/swift-transformers/pull/151) and [ml-explore/mlx-swift-examples#173](https://github.com/ml-explore/mlx-swift-examples/pull/173).

I'm really sorry, I've been quite busy with work lately. I will take some time over the weekend to review it. Thank you for your work.🍻

**Comment by DePasqualeOrg** [comment]
Honestly, no, I don't think that's a good idea. I spent several whole days trying to get as close as possible to the Python Jinja2. I think that should be our reference point, not the TypeScript version. That is what is used for chat templates in Python.

If you have tests from your implementation of the TypeScript version, you can share them and we can see if they pass on my branch. You can take my tests and see if they pass on your branch.

It was an enormous amount of work to get this working, and I would be very disappointed if it gets thrown away.

**Comment by johnmai-dev** [comment]
I'm very sorry, but it won't be thrown away. I think the ported Jinja2 still needs further inspection and testing, so I want to split some of Jinja2's code into smaller modules for iterative updates. 

Secondly, the original intention of Swift Jinja is to replicate TypeScript Jinja; I hope to keep it synchronized with the TypeScript Jinja version and then port Jinja2 on this basis. 

Replicating first is the fastest method, and there are significant design differences between Swift Jinja and Jinja2. If we truly want to replicate Jinja2, a refactor of Swift Jinja might be necessary.

**Comment by DePasqualeOrg** [comment]
I really don't know what you have in mind. What exactly do you want to separate out from my contribution? Is there any functionality in your branch that isn't reproduced in mine?

Please also point out specific differences between the TypeScript version and the Python Jinja2 that you think are important.

My pull request has been open for more than a month, and only now am I learning about your parallel effort, which you didn't share here. Since no progress has been made for several months, I decided to take the initiative and make chat templates for vision models and function calling work. Now they're working.

If you're worried about correctness, I can remove the filters for which I disabled tests because fully implementing them would be too complex.

However, considering the huge effort required to make this all work, I don't think we're going to be able to merge your version and mine. Since I was the first to share this here, I think we should use my branch as the basis for further work. I'm happy to remove anything that you think is not ready for production, and to add any functionality from your branch that is missing in mine.

**Comment by DePasqualeOrg** [comment]
I've reviewed my PR, and there are no major architectural changes here. Therefore, I don't understand what you mean about refactoring. Furthermore, I'd like to emphasize that I first ported functionality from the TypeScript implementation and then (after almost entirely covering the TypeScript implementation) added missing functionality (mainly the filters and tests) from Jinja2 in Python. This PR is still comparable to the TypeScript implementation.

I'm hoping we can move forward with this efficiently, because I want to start building actual features in apps using function calling and vision models instead of getting bogged down with this library.

**Comment by johnmai-dev** [comment]
Ok. 

Before merging this PR, you still need to resolve the previous review feedback. @DePasqualeOrg 

Additionally, I hope @pcuenca  can also participate in the review since swift-transformers is mainly being used at present.

File: Python/test-chat-template.ipynb
[the remark this answers is not in the document]
Comment (DePasqualeOrg) [reply Python/test-chat-template.ipynb:1]: Please read the first line of the notebook:

> This notebook can be used to generate output with Jinja2 in Python to compare against the output from Swift Jinja. Note that Jinja2 doesn't preserve the insertion order of keys in dictionaries, which may result in differences with Swift Jinja.

I believe it won't be included in the files downloaded by consumers of this package, because it is not included in Package.swift. It will be useful for creating more tests with chat templates as new models come out.

File: Tests/Templates/ToolSpecs.swift [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+//
+
+import OrderedCollections
+
+struct ToolSpec {
+    static let getCurrentWeather = OrderedDictionary(uniqueKeysWithValues: [
+        ("type", "function") as (String, Any),
+        (
+            "function",
+            OrderedDictionary(uniqueKeysWithValues: [
```
[the remark this answers is not in the document]
Comment (DePasqualeOrg) [reply Tests/Templates/ToolSpecs.swift:15]: I already discussed this in an earlier comment in this PR. This is necessary in order to maintain the order of the keys. Otherwise the test will not pass.

File: Sources/Parser.swift [hunk 2]
```diff
@@ -409,95 +471,120 @@ func parse(tokens: [Token]) throws -> Program {
...
-                "Expected identifier/tuple for the loop variable, got \(type(of:loopVariable)) instead"
+                "Expected identifier/tuple for the loop variable, got \(type(of: loopVariable)) instead"
             )
         }
-
         try expect(type: .in, error: "Expected `in` keyword following loop variable")
-
         let iterable = try parseExpression()
-
+        // Handle optional if condition for filtering
```
[the remark this answers is not in the document]
Comment (DePasqualeOrg) [reply Sources/Parser.swift:507]: Yes, it is used on line 533.

File: Tests/Templates/ToolSpecs.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (johnmai-dev) [reply Tests/Templates/ToolSpecs.swift:15]: Which specific test case?

File: Python/test-chat-template.ipynb
[the remark this answers is not in the document]
Comment (johnmai-dev) [reply Python/test-chat-template.ipynb:1]: The ipynb needs to be removed. If a comparison is needed, please do it locally.

File: Sources/Ast.swift [hunk 3]
```diff
@@ -59,14 +60,14 @@ struct Identifier: Expression {
...
-extension Identifier: Loopvar {}
-extension TupleLiteral: Loopvar {}
+typealias Loopvar = Expression
 
 struct For: Statement {
     var loopvar: Loopvar
     var iterable: Expression
     var body: [Statement]
+    var defaultBlock: [Statement]
+    var ifCondition: Expression?
```
[the remark this answers is not in the document]
Comment (DePasqualeOrg) [reply Sources/Ast.swift]: The TypeScript implementation appears to be incomplete. This is why it is helpful to use the original Jinja2: https://github.com/pallets/jinja/blob/6aeab5d1da0bc0793406d7b402693e779b6cca7a/src/jinja2/nodes.py#L319

I have renamed this `test` to better align with Jinja2. If I remember correctly, it was necessary in order to correctly interpret one of the chat templates for tools.

File: Python/test-chat-template.ipynb
[the remark this answers is not in the document]
Comment (DePasqualeOrg) [reply Python/test-chat-template.ipynb:1]: It doesn't need to be removed, but if you want it to be removed, I'll remove it.

File: Tests/Templates/ToolSpecs.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (DePasqualeOrg) [reply Tests/Templates/ToolSpecs.swift:15]: Most of the ones in which dictionaries are stringified in the template output. This is something that took a long time to resolve, and I can assure you that it is necessary.

File: Python/test-chat-template.ipynb
[the remark this answers is not in the document]
Comment (johnmai-dev) [reply Python/test-chat-template.ipynb:1]: Not everyone uses ipynb; I tend to use py more, and there's no need to install the large Jupyter environment. Also, not all functions require comparing the differences between Swift Jinja and Jinja2. I suggest doing the comparison work locally instead.

[the remark this answers is not in the document]
Comment (DePasqualeOrg) [reply Python/test-chat-template.ipynb:1]: It's just a notebook with no dependencies. It would have been entirely optional to use it, and perhaps useful to someone who wants to add more tests. I provided it for convenience, but now I've removed it.

File: Tests/Templates/ToolSpecs.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (johnmai-dev) [reply Tests/Templates/ToolSpecs.swift:15]: I approve of your approach. But it seems a bit complicated for Jinja users; is there a simpler way to use it?

[the remark this answers is not in the document]
Comment (DePasqualeOrg) [reply Tests/Templates/ToolSpecs.swift:15]: Jinja users don't need to use the OrderedDictionary. You can just use a native Swift dictionary.

**Review by johnmai-dev — APPROVED** [review]
Thank you very much @DePasqualeOrg! 
I have reviewed it and think it can be merged at any time. 

But we still need @pcuenca to help review it again.

**Comment by DePasqualeOrg** [comment]
@pcuenca, could you let us know if you want to review this? If you don't have time, I think we should merge it so that I can move ahead with https://github.com/huggingface/swift-transformers/pull/151, https://github.com/ml-explore/mlx-swift-examples/pull/174, and https://github.com/ml-explore/mlx-swift-examples/pull/173.

File: Package.resolved
Comment (pcuenca) [inline Package.resolved:1]: Is it necessary to commit this file?

**Review by pcuenca — APPROVED** [review]
I haven't had the chance to test it in depth, but I'm supportive of merging to unblock downstream uses, as long as the repo owner is happy with the changes. In my opinion, please follow the lead of @johnmai-dev here.

There appears to be a failing test case, but it seems related to dictionary ordering. I would suggest to write test cases in such a way that ordering does not impact results, but we can do that in a different PR.

File: Package.resolved
Comment (DePasqualeOrg) [reply Package.resolved:1]: I'm not sure. It was created after I added swift-collections. Are lock files usually included for dependencies in Swift packages?

Comment (pcuenca) [reply Package.resolved:1]: I don't think so, Xcode should resolve the dependencies when you compile. I'd suggest to remove.

It's also a bit unfortunate that another dependency had to be added, is it just to guarantee reproducibility for tests? If so, we should be able to move it to the test target (but we can do it later).

Comment (DePasqualeOrg) [reply Package.resolved:1]: swift-collections is necessary for the reproducibility of some tests, and is also needed in the library itself. I'm not sure how else you would test stringification of dictionaries if you don't have a deterministic key order.

I see that mlx-swift has a Package.resolved file, and Claude says it's usually included in Swift packages, for what it's worth.

File: Sources/Parser.swift
Comment (johnmai-dev) [inline Sources/Parser.swift:1]: Does this commit fix this issue? https://github.com/johnmai-dev/Jinja/issues/12

**Comment by johnmai-dev** [comment]
Are there any other commits that need to be pushed? If not, I think we can merge. @DePasqualeOrg

File: Package.resolved
Comment (johnmai-dev) [reply Package.resolved:1]: Apple recommend committing the Package.resolved.

https://developer.apple.com/documentation/xcode/adding-package-dependencies-to-your-app#Coordinate-package-versions-across-your-team

https://developer.apple.com/documentation/xcode/making-dependencies-available-to-xcode-cloud#Use-Swift-package-dependencies-and-Git-submodules

**Comment by awni** [comment]
🚀  this should unblock the Deep Seek reasoning chat templates so looking forward to getting it landed!

**Comment by DePasqualeOrg** [comment]
I have nothing more to add, so you can merge this, @johnmai-dev.

**Comment by johnmai-dev** [comment]
Thank you for your efforts! @DePasqualeOrg 
Jinja Swift 1.1.0 Released!

**Comment by alelordelo** [comment]
@DePasqualeOrg , thanks for this awesome contribution!

I am using FullMoon , a super clean swiftUI client, build with  MLX and Jinja.
https://github.com/mainframecomputer/


Can you give some instructions on how can we parse a function call with the new Jinja 1.1?

**Comment by DePasqualeOrg** [comment]
@alelordelo, when the model responds with the function call, it's up to the app to handle it. Jinja is not involved at all at that point.
